lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,237 @@
+        <defs>
+          <clipPath id='${clipId}' clipPathUnits='objectBoundingBox'>
+            <polygon points="0,0  1,0  1,${top} 0,${top}" ></polygon>
+            <polygon points="0,${bottom} 1,${bottom} 1,1 0,1" ></polygon>
+            <polygon points="0,${top} ${left},${top} ${left},${bottom} 0,${bottom}" ></polygon>
+            <polygon points="${right},${top} 1,${top} 1,${bottom} ${right},${bottom}" ></polygon>
+          </clipPath>
+        </defs>
+      </svg>`;
+    return clipPathSvg;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  static _installBackgroundImageStyle(dom, fullPageScreenshot) {
+    const containerEl = dom.find('.lh-container', dom._document);
+    if (containerEl.querySelector('#full-page-screenshot-style')) return;
+
+    const fullpageScreenshotUrl = fullPageScreenshot.data;
+    const fullPageScreenshotStyleTag = dom.createElement('style');
+    fullPageScreenshotStyleTag.id = 'full-page-screenshot-style';
+    fullPageScreenshotStyleTag.innerText = `
+      .lh-element-screenshot__image {
+        background-image: url(${fullpageScreenshotUrl})
+      }`;
+    containerEl.appendChild(fullPageScreenshotStyleTag);
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {ParentNode} templateContext
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   */
+  static installOverlayFeature(dom, templateContext, fullPageScreenshot) {
+    ElementScreenshotRenderer._installBackgroundImageStyle(dom, fullPageScreenshot);
+
+    for (const el of dom.document().querySelectorAll('.lh-element-screenshot')) {
+      el.addEventListener('click', () => {
+        const overlay = dom.createElement('div');
+        overlay.classList.add('lh-element-screenshot__overlay');
+        const clipRect = {
+          width: Number(el.getAttribute('rectWidth')),
+          height: Number(el.getAttribute('rectHeight')),
+          left: Number(el.getAttribute('rectLeft')),
+          right: Number(el.getAttribute('rectRight')),
+          top: Number(el.getAttribute('rectTop')),
+          bottom: Number(el.getAttribute('rectBottom')),
+        };
+        overlay.appendChild(ElementScreenshotRenderer.render(
+          dom,
+          templateContext,
+          clipRect,
+          fullPageScreenshot,
+          {
+            // TODO: should this be documentElement width?
+            width: window.innerWidth * 0.75,
+            height: window.innerHeight * 0.75,
+          }
+        ));
+        document.body.appendChild(overlay);
+        overlay.addEventListener('click', () => {
+          overlay.remove();
+        });
+      });
+    }
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {ParentNode} templateContext
+   * @param {LH.Artifacts.Rect} clipRect Region of screenshot to highlight.
+   * @param {LH.Artifacts.FullPageScreenshot} fullPageScreenshot
+   * @param {Size} viewportSize Size of region to render screenshot in.
```

Comment:
```suggestion
   * @param {Size} containerSize Size of region to render screenshot in.
```

or...
`maxScreenshotSize`
`visibleArea`
`availableArea`

viewport makes me think of something else :)

Reply:
All those seem too generic. What's wrong with viewport?

Reply:
We have 1018 usages of "viewport" across 122 files thusfar and AFAICT *all* of them mean the viewport of the page :)

![image](https://user-images.githubusercontent.com/2301202/85326579-227c8700-b493-11ea-8976-27ee360e6675.png)

Understandably when I see a parameter or variable called `viewportSize` I would expect a developer unfamiliar with this specific function (or honestly any web developer not familiar with non-web computer graphics) to think it means the size of the page's viewport, but in this case it doesn't. Even if it perfectly described what was happening in computer graphics terminology it would still be worthy of clarification for this reason alone.

> All those seem too generic

I'm glad you share my concern about the vagueness of these terms :) I'd prefer to get *much* more specific anyhow 👍 

We have a bunch of different sizes floating around, so let's get as descriptive as possible.  Suffix anything in screenshot coords with `InScreenshotCoords` and anything in display coords with `InDisplayCoords`, keeping in separate objects to avoid suffixes or using similar prefixes or abbreviations all sound fine to me. Just looking for clarity.

*Known Facts As Input*
- The size, in display scale, of the container that this element screenshot is being rendered into. Currently `viewportSize`, my proposal `renderContainerSizeInDisplayCoords`
- The size, in screenshot scale, of the element in the full page screenshot. Currently `clipRect`,  my proposal `elementRectInScreenshotCoords`

*Choices*
- The desired size, in screenshot scale, of the target element preview in the full page screenshot. (includes our desire to include additional context about where the element was). Currently unnamed, my proposal `targetElementPreviewRectInScreenshotCoords`

*To Be Computed/Derived*
- The ratio of the display scale to screenshot scale of the element preview. Currently `zoomFactor`/`zoomRatio`, my proposal extract the computation to a function and keep `zoomFactor` here, hide internals of `zoomRatio` and rename to `targetZoomFactor`
- The size, in display scale, of the element screenshot's viewport into the underlying full page screenshot. Currently unnamed, my proposal `imageViewportInDisplayCoords`
- A bunch of intermediate stuff in between that isn't currently named required to translate `targetElementPreviewRectInScreenshotCoords` into something that fits into `imageViewportInDisplayCoords`

Open to name suggestions